src/Microsoft.AspNet.Hosting/Internal/HostingLoggerExtensions.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.AspNet.Http;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNet.Hosting.Internal
+{
+    internal static class HostingLoggerExtensions
+    {
+        public static IDisposable RequestScope(this ILogger logger, HttpContext httpContext)
+        {
+            return logger.BeginScopeImpl(new HostingLogScope(httpContext));
+        }
+
+        public static void RequestStarting(this ILogger logger, HttpContext httpContext)
+        {
+            if (logger.IsEnabled(LogLevel.Information))
+            {
+                logger.Log(
+                    logLevel: LogLevel.Information,
+                    eventId: 1,
+                    state: new HostingRequestStarting(httpContext),
+                    exception: null,
+                    formatter: HostingRequestStarting.Callback);
+            }
+        }
+
+        public static void RequestFinished(this ILogger logger, HttpContext httpContext)
+        {
+            if (logger.IsEnabled(LogLevel.Information))
+            {
+                logger.Log(
+                    logLevel: LogLevel.Information,
+                    eventId: 2,
+                    state: new HostingRequestFinished(httpContext),
+                    exception: null,
+                    formatter: HostingRequestFinished.Callback);
+            }
+        }
+
+        private class HostingLogScope : ILogValues
+        {
+            private readonly HttpContext _httpContext;
```

Comment:
It turns out it was a bug to check the IsEnabled to short-circuit the scope. Even if messages in the `logger`'s category are entirely disabled, there can be other messages in the application that will be logged and will need the logger scope available... We might be able to optimize the BeginScopeImpl a little bit more and get the per-request overhead down to a single object with one field, then measure it from there to see how much impact that is having